### ISSUE TITLE
pgwire: support GSSAPI/Kerberos authentication

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#32526]
PM: @rolandcrosby 

tldr: Implement server-side authentication (Authn), not Authorization (Authz) using GSSAPI for pgwire sessions. **Try to avoid re-implementing the wheel in Go.** (see comment by Roland below)

From Roland:

> For that matter, we should investigate whether it's even worth using an existing Go wrapper for GSSAPI in the first place. There's a lot of functionality defined in GSSAPI, and we only need a small part of it (on the server side, Postgres does almost all of its work in [this code block](https://github.com/postgres/postgres/blob/master/src/backend/libpq/auth.c#L1089-L1316)). Writing our own interface that just wraps the functions/constants/structures we need might be the more prudent approach.

## Product Goals

> The short version is: build Kerberos-based authentication via GSSAPI on the server side, ensure that these users can be integrated with CockroachDB's existing enterprise RBAC features for authorization, and push back on requests for LDAP-based authorization.

Long version:

- We should support Kerberos client authentication on CockroachDB servers, using GSSAPI, in much the same way that PostgreSQL supports this authentication mechanism.
This will require CockroachDB either to automatically create users upon validation of GSSAPI authentication data, or to set up some kind of mapping between manually-created CockroachDB users and GSSAPI-authenticated users.
- If we have the bandwidth to do it, it would be nice to update the `cockroach sql` client to support Kerberos/GSSAPI authentication, but this is not essential for customers.
Note: we will need some client support to implement our own tests for the thing above
TBD whether we should attempt to support in-browser GSSAPI SPNEGO authentication for the admin UI; it would be very cool to demo but doesn’t seem to be a core requirement for customers. (Need to align with admin UI team.)
- We should not commit to integration with external authorization systems in the 2.2 timeframe - instead, we should focus on ensuring roles can be manually assigned to external users with the existing RBAC system.

## Current State: Authentication

For SQL connections, CockroachDB currently supports three authentication mechanisms, all of which are also supported by the `cockroach sql` client:

- none (insecure mode) - the user supplies a username, and is granted access as long as the username already exists in the CockroachDB system.users table
- password - the user supplies a username and password; the password is hashed, and the user is granted access if the username and hashed password exist in the system.users table
- certificate - the user supplies a username and X.509 certificate; the certificate’s validity is checked against the certificate authorities configured within CockroachDB, and the user is granted access if the common name in the certificate matches an entry in the system.users table

For web UI access, CockroachDB supports two authentication mechanisms:

- none (insecure mode) - anyone who can connect to the cluster’s HTTP(S) endpoints can access the web UI
- password - the user supplies a username and password via a web form; if the user exists in the system.users table and the hashed password matches, the user can access the web UI
(Certificate authentication support was not built for the web UI because the user experience for adding client certificates to most web browsers was determined to be difficult and error-prone.)

## Current State: Authorization

CockroachDB currently supports two authorization mechanisms:

- SQL grants - explicitly assign a set of privileges (SELECT, INSERT, CREATE, etc.) to a user on a per-database or per-table basis
- role-based access control - users can be assigned to roles, which are sets of privileges managed as a group

CockroachDB does not have any facilities for externally-managed authorization at this point.

## Customer Problems with Current State

Enterprise customers typically have an existing identity management system in place and are loath to deploy new systems that do not integrate with it. With a central identity management system, creation, management, and revocation of human and service accounts can be managed in a single place.

Although CockroachDB’s existing certificate-based authentication system can be used with identity management products that integrate with a certificate authority, there are some shortcomings that make this a less-than-ideal choice. Certificates generally have long expiration periods and are difficult or impossible to revoke once issued, and CockroachDB currently does not support certificate revocation at all.

What we’ve specifically heard from customer, our primary design partner for this project, is:
- They want centralized user management, managed by their existing Active Directory system.
- Some of their services currently use LDAP for authentication, but they are actively moving away from this in favor of Kerberos.
- They would like to manage database authorization centrally by assigning users to groups in Active Directory, with some kind of mapping between CockroachDB RBAC roles and Active Directory user groups

This seems more aspirational at the moment; they have not actually set this up with any of their existing database systems.

The most important authorization scenario is service-to-database authentication; managing human users is a secondary concern.

### Active Directory

Active Directory is Microsoft’s identity management and directory service. It is ubiquitous in enterprises; to date, it is the only specific identity management system that I’ve seen customers ask us to integrate with. Fortunately, building “Active Directory integration” doesn’t necessarily involve any proprietary protocols or work specific to Windows Server environments, since under the hood it uses LDAP and Kerberos to provide directory services and authentication.

### LDAP

Lightweight Directory Access Protocol (LDAP), as the name implies, is a protocol that enables users to query a structured directory of entities within an organization. LDAP directories typically contain information about human users, machine users, applications/services, and other pieces of IT infrastructure. Entities in directory services can be organized into hierarchies and assigned to groups.

Users in an LDAP directory can have passwords, and connections to an LDAP server can be authenticated using username/password authentication. Some systems, including PostgreSQL, can use LDAP for authentication. In this case, the user supplies a username and password to the service. The service attempts to authenticate to the LDAP system using the provided credentials; if the LDAP authentication operation succeeds, the service considers the user authenticated.

Using LDAP for service authentication like this is a suboptimal solution for a few reasons:
- It requires the user to provide their credentials directly to the service, rather than to a single central authentication server
- The service itself is not authenticated with the LDAP server - the service is essentially impersonating the user for the purposes of authenticating them

### Kerberos

Kerberos is a standard for authentication among clients and services in an environment with a central authorization server. Typically, human users start a Kerberos session by sending username and password credentials to the authentication server, while machine users are provided with a longer-lived Kerberos “keytab” file containing cryptographic keys.

User-to-service authentication in a Kerberos environment is mediated by the Kerberos server. When a user wants to authenticate to a “Kerberized” application, the user sends a request to the Kerberos server requesting a service ticket, which is an encrypted token specific to the user and service. The user then sends this service ticket to the service, which decrypts and validates the ticket using the keys stored in its keytab. If the service successfully decrypts the ticket, it considers the user authenticated.

### GSSAPI

While Kerberos is an open standard described by numerous RFCs, it does not specify a standard API - different Kerberos implementations provide their own APIs. GSSAPI is a generic authentication API (a set of C function prototypes, constants, and structures) that is used in practice to provide a standard API around Kerberos.

### SPNEGO

SPNEGO is a mechanism for negotiating available GSSAPI authentication options. HTTP servers and web browsers have implemented SPNEGO using the “WWW-Authenticate: Negotiate” HTTP header. This enables websites to natively support Kerberos/GSSAPI authentication - typically, either by prompting for a user’s credentials and starting a new Kerberos session, or for passwordless single sign-on by requesting a service ticket using the existing Kerberos session initiated when the user logged

**Comment by knz** [comment]
@mjibson one thing that would be nice once you have a clear(er) picture in your mind of the work to do, is to enumerate the sub-tasks as a comment to this issue. This way Roland will be better able to follow along.

**Comment by rolandcrosby** [comment]
Closed by #34772.
